snuba/subscriptions/entity_subscription.py

```diff
@@ -0,0 +1,122 @@
+from snuba.query.types import Condition
+
+
+class SubscriptionType(Enum):
+    LEGACY = "legacy"
+    SNQL = "snql"
+    DELEGATE = "delegate"
+
+
+class EntitySubscription(ABC):
+    def __init__(self, data_dict: Mapping[str, Any]) -> None:
+        ...
+
+    @abstractmethod
+    def get_entity_subscription_conditions_for_snql(
+        self, offset: Optional[int] = None
+    ) -> List[Expression]:
+        """
+        Returns a list of extra conditions that are entity specific and required for the
+        snql subscriptions
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_entity_subscription_conditions_for_legacy(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Condition]:
+        """
+        Returns a list of extra conditions that are entity specific and required for the
+        legacy subscriptions
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def to_dict(self) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+
+class SessionsSubscription(EntitySubscription):
+    def __init__(self, data_dict: Mapping[str, Any]) -> None:
+        super().__init__(data_dict)
+        try:
+            self.organization: int = data_dict["organization"]
+        except KeyError:
+            raise InvalidQueryException(
+                "organization param is required for any query over sessions entity"
+            )
+
+    def get_entity_subscription_conditions_for_snql(
+        self, offset: Optional[int] = None
+    ) -> List[Expression]:
+        return [
+            binary_condition(
+                ConditionFunctions.EQ,
+                Column(None, None, "org_id"),
+                Literal(None, self.organization),
+            ),
+        ]
+
+    def get_entity_subscription_conditions_for_legacy(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Condition]:
+        return []
+
+    def to_dict(self) -> Mapping[str, Any]:
+        return {"organization": self.organization}
+
+
+class BaseEventsSubscription(EntitySubscription, ABC):
+    def get_entity_subscription_conditions_for_snql(
+        self, offset: Optional[int] = None
+    ) -> List[Expression]:
+        if offset is None:
+            return []
+
+        return [
+            binary_condition(
+                ConditionFunctions.LTE,
+                FunctionCall(
+                    None, "ifNull", (Column(None, None, "offset"), Literal(None, 0)),
+                ),
+                Literal(None, offset),
+            )
+        ]
+
+    def get_entity_subscription_conditions_for_legacy(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Condition]:
+        if offset is None:
+            return []
+
+        return [[["ifNull", ["offset", 0]], "<=", offset]]
+
+    def to_dict(self) -> Mapping[str, Any]:
+        return {}
+
+
+class EventsSubscription(BaseEventsSubscription):
+    ...
```

Comment:
I am not familiar with ellipsis in this context. I've always seen this written as `pass`. Do you know what is the difference?

Reply:
Yeah its just a built in singleton object. A cuter way of writing `pass` imo
Its just that statement needs to be there after the function def if no logic will be defined